src/extension.ts

```diff
@@ -499,6 +500,72 @@ function launchMetals(
       });
     });
 
+    registerCommand("metals.new-scala-file", (dirArg: Uri) => {
+      const directory = (() => {
+        if (dirArg !== undefined) {
+          return dirArg.toString();
+        } else if (window.activeTextEditor) {
+          const currentPath = window.activeTextEditor.document.uri.fsPath;
+          return Uri.file(path.dirname(currentPath)).toString();
+        }
+        // NOTE(aleksei): I would return workspace folder here, but not sure if there's API for that.
+        // Moving this funcitonality to server.
+        return undefined;
+      })();
+
+      const classPick = { label: "Class", kind: "class" };
+      const objectPick = { label: "Object", kind: "object" };
+      const traitPick = { label: "Trait", kind: "trait" };
+      const packageObjectPick = {
+        label: "Package Object",
+        kind: "package-object"
+      };
+      const worksheetPick = { label: "Worksheet", kind: "worksheet" };
+
+      return window
+        .showQuickPick(
```

Comment:
Instead of implementing the quick pick here, I think it would be nice to add a new LSP extension so the Metals server can open a quick pick. We can use window/showMessageRequest as a fallback. 

I have tried to follow the principle of avoiding too much custom logic in the vscode client and allowing the server to push behavior instead. The benefit of this strategy is that we 1) we can reuse more code to support multiple editors and 2) next time we need a quick pick we may not have to update the client code.

Reply:
I have some comments about the implementation, but I too think we should move this to the server similarly to how we do for other extensions.

Reply:
@olafurpg , so this way part of the logic for the dialog about a new file (it's kind and name) is going to be transferred to the server and _inside_ of the new-scala-file command execution. Dialog parts can be cancelled, so I have a question - how this should be handled by the command? Currently it returns a string/URI of the newly created file. Would using null in case of cancellation be alright for this? Or the request should finish with failure? Or instead of just string as the return type, we can use some `Cancelable` interface / generic container with it inside?

Reply:
If the command “create new scala file” is sent to the server with a directory as the argument (or something like that) then I think it doesn’t matter what Metals returns. Metals already knows how to do the quick pick, do input box, apply workspace edit and also change the focused window (goto locations, it’s used in tree view protocol). Does that make sense?

Reply:
The directory argument can be optional, for example when triggered via the command palette. Metals already knows the focused document.